docs/src/notebooks/structured.jl

```diff
@@ -0,0 +1,229 @@
+ Symmetric(x2) |> dump
+
+# ╔═╡ dbe34880-93bf-4a5d-b28b-5e6b76267742
+ d_x2_sym |> dump
+
+# ╔═╡ 0122a4df-75d9-444e-8d83-d7a93b6dfeb5
+begin
+	struct MySymmetric2{T,S<:AbstractMatrix{<:T}} <: AbstractMatrix{T}
+	    data::S
+	    uplo::Char
+	
+	    function MySymmetric2{T,S}(data, uplo::Char) where {T,S<:AbstractMatrix{<:T}}
+	        LinearAlgebra.require_one_based_indexing(data)
+	        (uplo != 'U' && uplo != 'L') && LinearAlgebra.throw_uplo()
+	        new{T,S}(data, uplo)
+	    end
+	end
+	function MySymmetric2(A, uplo='U')
+		 MySymmetric2{eltype(A), typeof(A)}(A, 'U')
+	end
+end
+
+# ╔═╡ 8497709d-d123-48bc-a86e-5f58aa1b0ebc
+@inline function Base.getindex(A::MySymmetric2, i::Int, j::Int)
+    @boundscheck Base.checkbounds(A, i, j)
+    @inbounds if i == j
+        return A.data[i, j]
+    elseif (A.uplo == 'U') == (i < j)
+        return A.data[i, j]
+    else
+        return A.data[j, i]
+    end
+end
+
+# ╔═╡ b648f8e9-ec43-4883-acd1-3f77f85f23c8
+md"""
+Let us implement a Symmetric matrix type like it exist in LinearAlgebra.jl
+"""
+
+# ╔═╡ 470f5618-109e-4328-9e8c-21d130b0170c
+md"""
+!!! warning
+    The gradient of a Symmetric matrix seem to be right right! But let's check for the correctness of the off-diagonal entries, when they are non-zero!
+"""
+
+# ╔═╡ dbbc0d6c-dbab-4e8d-92cc-6c5517b7caac
+md"""
+So while the Symmetric matrix does actually store a full matrix underneath, it is more treated like a triangular matrix (upper in this case). So the entries `A[1,2]` and `A[2,1]` are aliased. The gradient accumulation is correct, but when we then use this matrix later on in Julia code it appears as if the gradient was twice what it ought to be.
+"""
+
+# ╔═╡ bd3ea593-88e2-4f6c-9df7-551950cdf020
+md"""
+!!! note
+    We implement a second equivalent type here, since otherwise we could not show the issue clearly.
+"""
+
+# ╔═╡ d0a031e4-99a4-417b-8f57-58a67219fa23
+begin
+	Base.size(A::MySymmetric2) = size(A.data)
+	Base.length(A::MySymmetric2) = length(A.data)
+end
+
+# ╔═╡ d047c162-446f-4de4-b2fd-5f2550f0ad78
+md"""
+Now we can implement a rule where we adjust the gradient contribution to be half since later we will double count them.
+"""
+
+# ╔═╡ e81ac66b-75ba-4e88-8e9c-491a60a671dc
+begin
+	function EnzymeRules.augmented_primal(config, func::Const{typeof(Base.getindex)}, ::Type{<:Active}, S::Duplicated{<:MySymmetric2}, i::Const, j::Const)
+	    # Compute primal
+	    if needs_primal(config)
+	        primal = func.val(S.val, i.val, j.val)
+	    else
+	        primal = nothing
+	    end
+	
+	    # Return an AugmentedReturn object with shadow = nothing
+	    return EnzymeRules.AugmentedReturn(primal, nothing, nothing)
+	end
+
+	function EnzymeRules.reverse(config, ::Const{typeof(Base.getindex)}, dret::Active, tape,
+	                 S::Duplicated{<:MySymmetric2}, i::Const, j::Const)
+		i = i.val
+		j = j.val
+		A = S.val
+		dA = S.dval
+		@inbounds if i == j
+        	dA.data[i, j] += dret.val
+    	elseif (A.uplo == 'U') == (i < j)
```

Comment:
I'm not sure about the correctness of this generally

Reply:
Yeah, which is why I haven't pushed this as a general rule for `Symmetric`, this notebook is more meant to explain to users what is happening.